### PR TITLE
ENHANCED: Version 2 of C++ interface

### PR DESCRIPTION
**Pull request by kamahen**
This still needs some "polish" but I think it's in good enough shape to commit.

Discussion about this is in https://swi-prolog.discourse.group/t/changes-to-swi-cpp-h/5601

This is an incompatible change -- when a new version of SWI-Prolog is created (probably 8.5.16),
also update code that uses SWI-cpp.h: packages-swipl-win, rocksdb, rolog

Additional comments from the commit message:
- removed some implicit conversions and constructors whose use can cause subtle bugs
- made most conversions and constructors explicit
- added subclass constructors to make Prolog types explicit and
  avoid some subtle bugs
- added copy constructors, assignment operators
- renamed handle_, ref_ to C_, for simplified "getter/setter"
- added PlCheck() for easier calling of SWI-Prolog.h functions
- removed conversion (cast) operators
- - added pseudo-exception PlFail for returning failure from a PREDICATE
  - also used by the unify_XXX_ex() methods
- deprecated assignment operator for term unification
  - replaced by unify_XXX() and unify_XXX_ex() methods
- added some support for std::string, std::wstring
  - and additional support for wchar_t
- added some more methods that allow using PlTerm:xxx(...) instead of
  PL_xxx(t.get_term_t(), ...); also for PlAtom
- added PlTerm::type(), PlTerm::is_variable(), etc. methods
- moved atom_t handle into private PlAtom, term_t into private PlTerm
  - added get_atom_t(), get_term_t()
  - added verify(), is_valid(), reset() methods
- some refactoring, e.g., PlTerm::check(), PlTerm::null
- refactored the integer "getters" to get around problems with C++
  built-in conversions and integer promotion.
- refactored comparison methods, added PlTerm::compare()
- replaced "int" as return type by "bool" in many places
- use common C++ naming practice for class fields (suffixed by "_")
- added unit tests
- moved examples from test.cpp to unit tests file (cpp4pl.cpp, test_cpp.pl)
- used C++ initializer lists where possible
- added [[nodiscard]] to many methods
- added [[deprecated]] to some methods
- (internal: used a template class to handle some boilerplate)

### CODE REVIEW

**Comment by JanWielemaker** [comment]
This pull request has been mentioned on **SWI-Prolog**. There might be relevant details there:

https://swi-prolog.discourse.group/t/changes-to-swi-cpp-h/5601/34

File: SWI-cpp.h [hunk 1]
```diff
@@ -62,69 +60,184 @@
...
-  int operator ==(functor_t to) const
-  { return functor == to;
+  size_t arity_sz() const {
+    return PL_functor_arity_sz(functor_);
+  }
+
+  operator bool() const = delete; // Use is_valid() instead
+
+  bool is_valid() const
+  { return functor_ != null && PL_functor_name(functor_) != 0;
```
Comment (kamahen) [inline SWI-cpp.h]: return functor != null && name().is_valid()

File: SWI-cpp.h [hunk 2]
```diff
@@ -62,69 +60,184 @@
...
+  }
+
+  void reset()
```
Comment (kamahen) [inline SWI-cpp.h]: Add: reset(functor_t)
- also PlAtom::reset(atom_t), PlTerm::reset(term_t)

File: SWI-cpp.h [hunk 3]
```diff
@@ -62,69 +60,184 @@
...
+
+  operator bool() const = delete; // Use is_valid() instead
+
+  bool is_valid() const
+  { return handle_ != null;
   }
-  PlAtom(const char *text)
-  { handle = PL_new_atom(text);
+
+  void reset()
```
Comment (kamahen) [inline SWI-cpp.h]: Add: reset(atom_t)

File: SWI-cpp.h [hunk 4]
```diff
@@ -62,69 +60,184 @@
...
+  { handle_ = null;
   }
-  PlAtom(const wchar_t *text)
-  { handle = PL_new_atom_wchars(wcslen(text), text);
+
+  atom_t get_atom_t() const { return handle_; }
+  explicit operator const atom_t() const { return get_atom_t(); }
+
+  [[deprecated]] operator const char *()    const { return c_str(); }
```
Comment (kamahen) [inline SWI-cpp.h]: [[deprecated("use c_str()")]]

File: SWI-cpp.h [hunk 5]
```diff
@@ -62,69 +60,184 @@
...
+  [[deprecated]] operator const wchar_t *() const { return wc_str(); }
+  [[deprecated]] operator std::string()     const { return string(); }
+  [[deprecated]] operator std::wstring()    const { return wstring(); }
+
+  const char* c_str() const
+  { return PL_atom_chars(get_atom_t());
```
Comment (kamahen) [inline SWI-cpp.h]: use handle_ instead of get_atom_t()

File: SWI-cpp.h [hunk 6]
```diff
@@ -62,69 +60,184 @@
...
+  }
+  const wchar_t *wc_str() const
+  { return PL_atom_wchars(handle_, nullptr);
+  }
+  const std::string string() const
+  { size_t len;
+    const char *s = PL_atom_nchars(get_atom_t(), &len);
```
Comment (kamahen) [inline SWI-cpp.h]: use handle_ instead of get_atom_t (and elsewhere)

File: SWI-cpp.h [hunk 7]
```diff
@@ -62,69 +60,184 @@
...
-  { return PL_atom_wchars(handle, NULL);
+  [[nodiscard]] bool operator ==(const std::wstring& s) const
+  { size_t len;
+    const wchar_t* s0 = PL_atom_wchars(handle_, &len);
+    return std::wstring(s0, len) == s;
+  }
+  [[nodiscard]] bool operator ==(const PlAtom &a) const
+  { return handle_ == a.handle_;
+  }
+  [[nodiscard]] bool operator ==(atom_t to) const
```
Comment (kamahen) [inline SWI-cpp.h]: [[deprecate("use PlAtom")]]

File: SWI-cpp.h [hunk 8]
```diff
@@ -62,69 +60,184 @@
...
+  { return !(*this == s);
   }
-  int operator ==(const wchar_t *s) const
-  { return wcscmp(s, PL_atom_wchars(handle, NULL)) == 0;
+  [[nodiscard]] bool operator !=(const PlAtom &a) const
+  { return !(*this == a);
   }
-  int operator ==(const PlAtom &a) const
-  { return handle == a.handle;
+  [[nodiscard]] bool operator !=(atom_t to) const
```
Comment (kamahen) [inline SWI-cpp.h]: deprecate

File: SWI-cpp.h [hunk 9]
```diff
@@ -135,87 +248,281 @@ class PlAtom
...
+  [[nodiscard]] explicit operator int()             const { return get_int(); }
+  [[nodiscard]] explicit operator uint32_t()        const { return uint32_t(); }
+  // explicit operator uint64_t() const { return int64_t(); } // Conflicts with get_term_t()
+  // explicit operator size_t()   const { return size_t(); }  // Conflicts with get_term_t()
+  [[nodiscard]] explicit operator double()          const { return get_float(); }
+  [[nodiscard]] explicit operator void *()          const { return pointer(); }
+  [[nodiscard]] explicit operator PlAtom()          const { return atom(); }
+
+  // All the get_XXX() functions throw an exception if they fail
+  // (because of the wrong Prolog type. If you want to be safe, use
```
Comment (kamahen) [inline SWI-cpp.h]: typo (paren)

File: SWI-cpp.h [hunk 10]
```diff
@@ -135,87 +248,281 @@ class PlAtom
...
+  // is_XXX() first to verify the type.
+  [[nodiscard]] term_t          get_term_t() const { return ref_; }
+  [[nodiscard]] /* [[deprecated("c_str")]] */ const char *    c_str() const;
```
Comment (kamahen) [inline SWI-cpp.h]: restore deprecated

File: SWI-cpp.h [hunk 11]
```diff
@@ -135,87 +248,281 @@ class PlAtom
...
+  void integer(uint64_t *v) const { *v = get_uint64_t(); }
+  [[nodiscard]] bool            get_bool() const; // atoms 'true', 'false'
+  [[nodiscard]] double          get_float() const;
+  [[nodiscard]] double          get_double() const { return get_float(); }
+  [[nodiscard]] void *          pointer() const;
+
+  void nchars(size_t *len, char **s, unsigned int flags);
+
+  [[nodiscard]] PlAtom atom() const;
+  [[nodiscard]] bool get_atom(PlAtom *a) const
```
Comment (kamahen) [inline SWI-cpp.h]: rename to get_if_atom ?

File: SWI-cpp.h [hunk 12]
```diff
@@ -135,87 +248,281 @@ class PlAtom
...
-  operator PlAtom(void) const;
-  operator void *(void) const;
 
-  int type() const
-  { return PL_term_type(ref);
+					/* Compounds */
+  [[nodiscard]] PlTerm operator [](ARITY_T index) const;
+  [[nodiscard]] ARITY_T arity() const;
+  [[nodiscard]] const char *name() const;
+  bool name_arity(PlAtom *name, ARITY_T *arity) const;
```
Comment (kamahen) [inline SWI-cpp.h]: should be void and implementation throws exception

File: SWI-cpp.h [hunk 13]
```diff
@@ -641,570 +1179,577 @@ class PlRegister
...
   ARITY_T arity;
-
-  if ( PL_get_name_arity(ref, &name, &arity) )
+  if ( PL_get_name_arity(ref_, &name, &arity) )
     return PL_atom_chars(name);
+  throw PlTypeError("compound", *this);
+}
 
-  throw PlTypeError("compound", ref);
+inline bool
```
Comment (kamahen) [inline SWI-cpp.h]: throw PlTypeError("compound", *this)

File: SWI-cpp.h [hunk 14]
```diff
@@ -135,87 +248,281 @@ class PlAtom
...
+  /* unify_XXX_ex() throws PlInstantiationError if fails */
+  void unify_term_ex(const PlTerm& t2);		/* term */
+  void unify_atom_ex(const PlAtom& a);		/* atom */
+  void unify_atom_ex(const char *v);		/* atom */
+  void unify_atom_ex(const wchar_t *v);		/* atom */
+  void unify_atom_ex(const std::string& v);	/* atom */
+  void unify_atom_ex(const std::wstring& v);	/* atom */
+  void unify_list_codes_ex(const char* v);	/* atom */
+  void unify_list_chars_ex(const char* v);	/* atom */
+  void unify_int_ex(int32_t v);			/* integer */
```
Comment (kamahen) [inline SWI-cpp.h]: unify_integer_ex()

File: SWI-cpp.h [hunk 15]
```diff
@@ -135,87 +248,281 @@ class PlAtom
...
+
+  [[nodiscard]] bool unify_term(const PlTerm& t2) { return check(PL_unify(ref_, t2.get_term_t())); }
+  [[nodiscard]] bool unify_atom(const PlAtom& a)  { return check(PL_unify_atom(ref_, a.get_atom_t())); }
+  [[nodiscard]] bool unify_atom(const char *v)    { return check(PL_unify_atom_chars(ref_, v)); }
+  [[nodiscard]] bool unify_atom(const wchar_t *v) { return check(PL_unify_wchars(ref_, PL_ATOM, static_cast<size_t>(-1), v)); }
+  [[nodiscard]] bool unify_atom(const std::string& v) { return check(PL_unify_atom_nchars(ref_, v.size(), v.data())); }
+  [[nodiscard]] bool unify_atom(const std::wstring& v) { return check(PL_unify_wchars(ref_, PL_ATOM, v.size(), v.data())); }
+  [[nodiscard]] bool unify_list_codes(const char* v) { return PL_unify_list_codes(ref_, v); }
+  [[nodiscard]] bool unify_list_chars(const char* v) { return PL_unify_list_chars(ref_, v); }
+  [[nodiscard]] bool unify_int(int32_t v)  { return check(PL_unify_integer(ref_, v)); }
```
Comment (kamahen) [inline SWI-cpp.h]: unify_integer()

File: SWI-cpp.h [hunk 16]
```diff
@@ -135,87 +248,281 @@ class PlAtom
...
+  { return PlTerm(PL_exception(qid));
+  }
+
+  operator bool() const = delete; // Use is_valid() instead
+
+  bool is_valid() const
+  { return ref_ != null;
+  }
+
+  void reset()
```
Comment (kamahen) [inline SWI-cpp.h]: add: reset(term_t)

File: SWI-cpp.h [hunk 17]
```diff
@@ -135,87 +248,281 @@ class PlAtom
...
+};
+
+class PlTerm_term_t : public PlTerm
+{
+public:
+  explicit PlTerm_term_t(term_t t)
+    : PlTerm(t) {}
+};
+
+class PlTerm_int : public PlTerm
```
Comment (kamahen) [inline SWI-cpp.h]: PlTerm_integer

File: SWI-cpp.h [hunk 18]
```diff
@@ -225,33 +532,33 @@ class PlTerm
...
+      a0_(t0.get_term_t()) {}
+
+  term_t termv() const
+  { return a0_;
   }
-  PlTermv(size_t n, term_t t0)
-  { a0   = t0;
-    size = n;
+
+  term_t size() const
```
Comment (kamahen) [inline SWI-cpp.h]: size_t size()

File: SWI-cpp.h [hunk 19]
```diff
@@ -307,23 +618,27 @@ class PlCharList : public PlTerm
...
+  // want the overhead of virtual methods - we want PlTerm to be a
+  // thin wrapper on term_t.
+  [[nodiscard]] [[deprecated("c_str")]] operator const char *() const { return c_str(); }
+  [[nodiscard]] [[deprecated("wc_str")]] operator const wchar_t *() const { return wc_str(); }
+  [[nodiscard]] const char *c_str() const;
+  [[nodiscard]] const wchar_t *wc_str() const;
 
   foreign_t plThrow()
-  { return static_cast<foreign_t>(PL_raise_exception(ref));
+  { return static_cast<foreign_t>(PL_raise_exception(get_term_t()));
```
Comment (kamahen) [inline SWI-cpp.h]: ref_ instead of get_term_t()

File: SWI-cpp.h [hunk 20]
```diff
@@ -435,121 +749,347 @@ class PlResourceError : public PlException
...
+{ if ( !ref_ ||
+       !PL_put_float(ref_, val) )
+    throw PlResourceError();
 }
 
-__inline
-PlTerm::PlTerm(void *ptr)
-{ if ( !(ref = PL_new_term_ref()) ||
-       !PL_put_pointer(ref, ptr) )
+inline
```
Comment (kamahen) [inline SWI-cpp.h:835]: Move this up, with other PlTerm_atom() constructors

File: SWI-cpp.h [hunk 21]
```diff
@@ -435,121 +749,347 @@ class PlResourceError : public PlException
...
+PlTerm_atom::PlTerm_atom(const PlAtom& a)
+  : PlTerm(PL_new_term_ref())
+{ if ( !ref_ )
     throw PlResourceError();
+
+  PL_put_atom(ref_, a.get_atom_t()); // always returns TRUE
+}
+
+inline
+PlTerm_atom::PlTerm_atom(atom_t a)
```
Comment (kamahen) [inline SWI-cpp.h]: Move this up, with other PlTerm_atom() constructors

File: SWI-cpp.h [hunk 22]
```diff
@@ -435,121 +749,347 @@ class PlResourceError : public PlException
...
+
+		 /*******************************
+		 *	    TERM (BODY)		*
+		 *******************************/
+
+					/* PlTerm --> C */
+
+inline const char *
+PlTerm::c_str() const
+{ char *s;
```
Comment (kamahen) [inline SWI-cpp.h:862]: const char * and use const_cast<> in call

**Review by kamahen — COMMENTED** [review]
I've been reading this with a different comparison tool, so I've seen a few things that next fixing.
Currently at line 1100 ... will continue my review tomorrow.

File: SWI-cpp.h [hunk 23]
```diff
@@ -62,69 +60,184 @@
...
+    : functor_(PL_new_functor(PL_new_atom_wchars(wcslen(name), name), arity))
+  { verify();
   }
-  PlFunctor(const wchar_t *name, ARITY_T arity)
-  { functor = PL_new_functor(PL_new_atom_wchars(wcslen(name), name), arity);
+
+  PlFunctor(const PlFunctor&) = default;
+  PlFunctor& operator = (const PlFunctor&) = default;
+
+  functor_t get_functor_t() const { return functor_; }
```
Comment (kamahen) [inline SWI-cpp.h]: rename get_functor_t() to c() here and elsewhere?
[@JanWielemaker  's suggestion]

File: SWI-cpp.h [hunk 24]
```diff
@@ -62,69 +60,184 @@
...
-  PlAtom(const char *text)
-  { handle = PL_new_atom(text);
+
+  void reset()
+  { handle_ = null;
   }
-  PlAtom(const wchar_t *text)
-  { handle = PL_new_atom_wchars(wcslen(text), text);
+
+  atom_t get_atom_t() const { return handle_; }
```
Comment (kamahen) [inline SWI-cpp.h]: Rename get_atom_t() to c() [and elsewhere]
@JanWielemaker 's suggestion

File: SWI-cpp.h [hunk 25]
```diff
@@ -62,69 +60,184 @@
...
   }
-  int operator ==(const PlAtom &a) const
-  { return handle == a.handle;
+  [[nodiscard]] bool operator !=(atom_t to) const
+  { return !(*this == to);
   }
-  int operator ==(atom_t to) const
-  { return handle == to;
+
+  void register_atom() const
```
Comment (kamahen) [inline SWI-cpp.h]: Rename `register_atom()` as `register()`. Also, `unregister()`.

File: SWI-cpp.h [hunk 26]
```diff
@@ -435,121 +749,347 @@ class PlResourceError : public PlException
...
+PlTerm::wstring() const
+{ wchar_t *s;
+  size_t len;
+  if ( PL_get_wchars(ref_, &len, &s, CVT_ALL|CVT_WRITEQ|BUF_STACK) )
+    return std::wstring(s, len);
+  throw PlTypeError("text", *this);
+}
+
+inline long
+PlTerm::get_long() const
```
Comment (kamahen) [inline SWI-cpp.h:895]: (Comment by @JanWielemaker )

This should call PL_get_long_ex() (and all similar cases). The reason is
that this function throws a representation error if the Prolog term is
an integer, but too large for a C long.  The overall idea is to have
the C interface generate the exceptions where possible.
```
inline long
PlTerm::get_long() const
{ long v;
   if ( PL_get_long(ref_, &v) )
     return v;
   throw PlTypeError("integer", *this);
}
```
-->
```
inline long
PlTerm::get_long() const
{ long v;

   if ( PL_get_long_ex(ref_, &v) )
     return c;

   check(false); // should always throw and exception, so me a derived
method.
}
```
Where
```
inline bool
PlTerm::check(int rc)
{ term_t ex;
   if ( !rc && (ex=PL_exception(0)) )
     throw PlResourceError(PlTerm(ex));
   return rc;
}
```
Should be
```
inline bool
PlTerm::check(int rc)
{ term_t ex;
   if ( !rc && (ex=PL_exception(0)) )
     PlException(PlTerm_term_t(ex).cppThrow());
   return rc;
}
```

File: SWI-cpp.h [hunk 27]
```diff
@@ -641,570 +1179,577 @@ class PlRegister
...
 }
 
-__inline int
+inline int
 PlCall(const wchar_t *goal)
 { PlQuery q("call", PlTermv(PlCompound(goal)));
   return q.next_solution();
 }
 
-
```
Comment (kamahen) [inline SWI-cpp.h:725]: The PlAtom::PlAtom constructors should be moved here, to minimize diffs.

**Review by kamahen — COMMENTED** [review]
Incorporated Jan's comments from a private email, plus a few more comments from me.

There is additional discussion at https://swi-prolog.discourse.group/t/changes-to-swi-cpp-h/5601/35

File: SWI-cpp.h (hunk 25 above)
Comment (kamahen) [reply SWI-cpp.h]: "register" is a reserved word in C++ (and C), so changing this to register_ref()

**Comment by kamahen** [comment]
The documentation file (`pl2cpp.doc`) hasn't been updated with the latest changes -- I'm waiting for any reactions to the source code changes and/or the Discourse discussion before finishing the documentation.

**Comment by kamahen** [comment]
My latest update simplifies some code by using a template class. No change in functionality.

**Comment by JanWielemaker** [comment]
This pull request has been mentioned on **SWI-Prolog**. There might be relevant details there:

https://swi-prolog.discourse.group/t/ffi-sdl-problem-in-c/5697/3

**Comment by JanWielemaker** [comment]
Thanks!   Merged as agreed by mail.   Didn't touch the docs, except for a non-working
reference to sections in the main manual.    Please base further PRs on the current master.

**Pull request closed without merging**